### DISCUSSION TITLE
Testing file downloads

### DISCUSSION THREAD

**Discussion by ohnotnow** [livewire/livewire#4205] [1. Help] [answered]
There's nothing leaping out at me in the testing docs, apologies if I've overlooked something.  I've got a component that has :
```php
    public function exportCsv()
    {
        return response()->streamDownload(function () {
            echo "Person,Type,Asset,Building,Room,Allocated\n";
            foreach ($this->assets as $asset) {
                echo $asset->owner->full_name . ',';
                echo .....;
            }
        }, 'recent-allocations.csv', ['Content-Type' => 'text/csv']);
    }
```
And I was hoping to be able to have a test along the lines of :
```php
        Livewire::actingAs($user)
            ->test(RecentReport::class)
            ->call('exportCsv')
            ->assertHeader('Content-Type', 'text/csv')
            ->assertHeader('Content-Disposition', 'attachment; filename="recent-allocations.csv"');
```
or something like :
```php
   ->assertPayloadSet('effects.download.name', 'recent-allocations.csv');
```
Just to cover that at least a download is triggered.  The first method fails as the response is obviously a livewire one not a regular browser download.  The second always gives me null (and I've tried just `download.name` or the like) so I'm guessing that's not the way to use the assertion?  The code for it seems to always look inside `serverMemo` which isn't where the download object is.

Does anyone have any pointers how I could check for the download being triggered?

**Comment by ralphjsmit** [comment, marked as answer]
Hey there, this is possible to test, though you need Laravel Dusk for that.

Livewire itself also[ has tests](https://github.com/livewire/livewire/blob/master/tests/Browser/FileDownloads/Test.php) very similar to what you want.

This is what you need:
```php
   /** @test */
    public function trigger_downloads_from_event_listener()
    {
        $this->onlyRunOnChrome();

        $this->browse(function ($browser) {
            Livewire::visit($browser, Component::class)
                ->waitForLivewire()->click('@emit-download') // ADAPT THIS LINE 
                ->waitUsing(5, 75, function () {
                    return Storage::disk('dusk-downloads')->exists('download-target.txt');
                });

            $this->assertStringContainsString(
                'I\'m the file you should download.',
                Storage::disk('dusk-downloads')->get('download-target.txt')
            );
        });
    }
```

**Reply by ohnotnow** [reply]
Ahhh - ok.  I was hoping to avoid having dusk tests for this little app - but if that's the way then I guess I'll go down this route.  Thanks!

**Reply by ralphjsmit** [reply]
Yeah, I'd feel the same way as you ;-) Unfortunately I don't see any other, so I guess that we'll need to accept a little overhead here🙃

**Comment by joshhanley** [comment]
@ohnotnow I just had a look in my code base, and I have a test to ensure a file gets downloaded, I'll share it below, and hopefully gives you ideas of how you might be able to achieve it.

This is the test (calling "pdf" action downloads the document as a PDF):

```php
$response = Livewire::test(DocumentShowPage::class, ['document' => $document])
    ->call('pdf');

$this->assertStringContainsString('download', $response->lastResponse->getContent());
```

You might be able to get more data off the last response object to test for specific details, but I haven't checked that.

But basically when a file gets downloaded using Livewire, the file and its details as sent to the front end in the response, and the front end triggers the download. So in theory you should be able to see all the details in the request.

Hope this helps!

**Reply by ralphjsmit** [reply]
Thanks for sharing this Josh! This looks like an even better way, would it support multiple downloads at once as well?

**Comment by jonneroelofs** [comment]
Try ->assertFileDownloaded($filename)
I added it to the reference page of the docs some time ago because I had also been searching for a while how to test downloads.

**Reply by joshhanley** [reply]
@jonneroelofs thanks! I just saw that used in PR #4191 and was going to suggest it. Much nicer! 😁

Any chance you could also PR that to the testing page of the docs?

Thanks!

**Reply by jonneroelofs** [reply]
Sure. Will try to look into it this week.

**Reply by joshhanley** [reply]
Great thanks 🙂
